internal/pkg/consul/client.go

```diff
@@ -0,0 +1,385 @@
+		ServiceID: client.serviceId,
+		AgentServiceCheck: consulapi.AgentServiceCheck{
+			HTTP:     client.healthCheckUrl,
+			Interval: client.healthCheckInterval,
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Checks to see if Consul contains the service's configuration.
+func (client *consulClient) HasConfiguration() (bool, error) {
+	if stemKeys, _, err := client.consulClient.KV().Keys(client.configBasePath, "", nil); err != nil {
+		return false, fmt.Errorf("checking configuration existence from Consul failed: %v", err)
+	} else if len(stemKeys) == 0 {
+		return false, nil
+	} else {
+		return true, nil
+	}
+}
+
+// Puts a full toml configuration into Consul
+func (client *consulClient) PutConfigurationToml(configuration *toml.Tree, overwrite bool) error {
+
+	configurationMap := configuration.ToMap()
+	keyValues := convertInterfaceToConsulPairs("", configurationMap)
+
+	// Put config properties into Consul.
+	for _, keyValue := range keyValues {
+		exists, _ := client.ConfigurationValueExists(keyValue.Key)
+		if !exists || overwrite {
+			if err := client.PutConfigurationValue(keyValue.Key, []byte(keyValue.Value)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Puts a full configuration struct into the Registry
+func (client *consulClient) PutConfiguration(configuration interface{}, overwrite bool) error {
+	bytes, err := toml.Marshal(configuration)
+	if err != nil {
+		return err
+	}
+
+	tree, err := toml.LoadBytes(bytes)
+	if err != nil {
+		return err
+	}
+
+	err = client.PutConfigurationToml(tree, overwrite)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// Gets the full configuration from Consul into the target configuration struct.
+// Passed in struct is only a reference for decoder, empty struct is ok
+// Returns the configuration in the target struct as interface{}, which caller must cast
+func (client *consulClient) GetConfiguration(configStruct interface{}) (interface{}, error) {
+	var err error
+	var configuration interface{}
+
+	exists, err := client.HasConfiguration()
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("the Registry (Consul) doesn't contain configuration for %s. Expected data at %s", client.serviceId, client.configBasePath)
+	}
+
+	// Update configuration data from Consul using decoder
+	updateChannel := make(chan interface{})
+	errorChannel := make(chan error)
+
+	decoder := client.newConsulDecoder()
+	decoder.Consul = client.consulConfig
+	decoder.Target = configStruct
+	decoder.Prefix = client.configBasePath
+	decoder.ErrCh = errorChannel
+	decoder.UpdateCh = updateChannel
+
+	defer decoder.Close()
+	defer close(updateChannel)
+	defer close(errorChannel)
+	go decoder.Run()
+
+	select {
+	case <-time.After(5 * time.Second):
+		err = errors.New("timeout loading config from client")
+	case ex := <-errorChannel:
+		err = errors.New(ex.Error())
+	case raw := <-updateChannel:
+		configuration = raw
```

Comment:
It looks like there could be lost functionality here. See the link:
https://github.com/edgexfoundry/edgex-go/blob/1c12c2fe67b376e4e8d590974ed9fd1a5305ddf7/internal/core/data/init.go#L191

Note the type checking and then the verification that the struct returned from Consul isn't empty. Is the expectation that the caller would still perform these checks?

Reply:
That logic remains in the calling services. See the example in the README.md